networking_nsxv3/plugins/ml2/drivers/nsxv3/agent/provider_nsx_mgmt.py

```diff
@@ -0,0 +1,880 @@
+            "resource_type": "SpoofGuardSwitchingProfile",
+            "white_list_providers": ["LPORT_BINDINGS"],
+            "display_name": os_id
+        }
+    
+    def network(self, os_net, provider_net):
+        return {
+            "resource_type": "LogicalSwitch",
+            "vlan": os_net.get("segmentation_id"),
+            "transport_zone_id": provider_net.get("transport_zone_id"),
+            "address_bindings": [],
+            "admin_state": "UP",
+            "description": "",
+            "display_name": os_net.get("id"),
+            "hybrid": False,
+            "switching_profile_ids": [],
+        }
+    
+    def qos(self, os_qos, provider_qos):
+        payload = {
+            "resource_type": "QosSwitchingProfile",
+            "display_name": os_qos.get("id"),
+            "tags": self.tags(os_qos),
+            "shaper_configuration": [],
+            "dscp": { "mode": "TRUSTED", "priority": 0 }
+        }
+
+        type = {"ingress": "IngressRateShaper", "egress": "EgressRateShaper"}
+
+        for rule in os_qos.get("rules"):
+            if "dscp_mark" in rule:
+                payload["dscp"] = {
+                    "mode": "UNTRUSTED",
+                    "priority": int(rule["dscp_mark"])
+                }
+                continue
+            payload["shaper_configuration"].append({
+                "resource_type": type.get(rule.get("direction")),
+                "enabled": True,
+                "average_bandwidth_mbps": \
+                    int(round(float(rule["max_kbps"]) / 1024)),
+                "peak_bandwidth_mbps": \
+                    int(round(float(rule["max_kbps"]) / 1024) * 2),
+                "burst_size_bytes": int(rule["max_burst_kbps"]) * 128
+            })
+        return payload        
+    
+    def port(self, os_port, provider_port):
+        p = os_port
+        pp = provider_port
+
+        p_pid = pp.get("id")
+        p_ppid = pp.get("parent_id")
+        p_qid = pp.get("qos_policy_id")
+
+        if not p_pid and not p_ppid:
+            LOG.error("Port '%s' not found.", p.get("id"))
+            return
+
+        port = {
+            "display_name": os_port.get("id"),
+            "logical_switch_id": \
+                p.get("vif_details").get("nsx-logical-switch-id"),
+            "admin_state": "UP",
+            "switching_profile_ids": pp.get("switching_profile_ids"),
+            "address_bindings": p.get("address_bindings"),
+            "attachment": {
+                "attachment_type": "VIF",
+                "id": os_port.get("id"),
+                "context": {
+                    "resource_type": "VifAttachmentContext",
+                    "vif_type": "PARENT",
+                    "traffic_tag": \
+                        p.get("vif_details").get("segmentation_id")
+                }
+            },
+            "tags": self.tags(os_port, more={NSXV3_SECURITY_GROUP_SCOPE:os_id \
+                for os_id in p.get("security_groups")})
```

Comment:
Assigning multiple security-groups per port doesn't work. With a dictionary, the static KEY (NSXV3_SECURITY_GROUP_SCOPE=security_group) can only exist once. Therefor, it will be always replaced with the last security group id in the array.

Reply:
Fixed